### ISSUE TITLE
[TW-1191] write vim color profiles for pending.data and completed.data

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1226]
_David Patrick on 2009-06-16T16:51:22Z says:_ 

 subject says it all

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2009-06-16T16:51:22Z
Modified: 2014-02-09T02:39:08Z
```

**Comment by taskwarrior** [comment]
_John Florian on 2009-06-24T19:52:06Z says:_ 

 Hey all, give this a whirl and see what you think thus far.  I haven't committed any of this as I'm not sure where to put it.  Generally VIM ships all of this type of stuff with itself and I've never seen any evidence of a vim.d type of directory where other packages can extend its functionality.

For now, just untar this into your home directory or wherever you'd normally keep your .vim directory.  If you don't already have such a dir, that should be fine IIRC.  These link to VIM's standard colors so that if you've diddled those, you'll see that reflected here as well.

It's complete AFACT, but that's based purely on my own live data which may not utilize everything.  Should be enough to give you an idea though.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-06-24T21:20:51Z says:_ 

 John, that's great, and works well!  Thank you!

**Comment by taskwarrior** [comment]
_John Florian on 2009-06-24T22:04:58Z says:_ 

 Paul, how should we proceed with this then?  Do we want to hold up for a while in case the file format needs to change?  Do we want to have task just make this available for user's to install like this if they so chose or should we coordinate with Bram to get this shipped with future versions of vim?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-06-25T00:22:03Z says:_ 

 Good question.  First, the file format is not changing (for a while), so let's discount that.  I see three options, and they are not mutually exclusive:

1) Make the .vim files downloadable from this site.
2) Include the .vim files in 1.8.0+ packages and install them.
3) Petition Bram to include them.

I suggest that we do 1 and coincide that with the 1.8.0 beta.
I would defer to Fredde on item 2.
I think we should do 3, but first we have some other .vim files needed.  For example, the command "task 1 edit" throws you into an editor, and maybe we could get that colored too?

**Comment by taskwarrior** [comment]
_Federico Hernandez on 2009-06-25T03:49:58Z says:_ 

 > 1) Make the .vim files downloadable from this site.
> 2) Include the .vim files in 1.8.0+ packages and install them.

> I suggest that we do 1 and coincide that with the 1.8.0 beta.
> I would defer to Fredde on item 2.

We can do both 1) and 2). We already had ideas about an add-on download page (something like that) as well as include the files in the release tarball and packages. Though they will not be "installed" all the way down. They will end up in the doc <doc-dir>/examples (/usr/share/doc/task-<version>/examples/vim) which is the canonical form to have additional, non-essentials files like scripts, etc distributed with packages.

We need as well to include a license statement in files. John, would it be OK for you if I add the GPLv2+ information text to your files?

We might otherwise risk the review process for Ubuntu/Debian. As well as the current "inclusion" in Fedora.


> I think we should do 3, but first we have some other .vim files needed.  For example, the command "task 1 edit" throws you into an editor, and maybe we could get that colored too?
> 3) Petition Bram to include them.

That would be the only wy to have them "systemwide" available. It would be the wrong way to make the task package depend on the different vim packages and then "install" the files into the vim package owned dirs. It would as well be cumbersome to pass QA reviews for the packages if we do stuff in user's homedirectories while let's say installing the rpm or deb. The user's that are interested in this functionality often know how to "install" these files manually anyhow. 

By having the vim task files included in the upstream project they will eventually dripple down into the different vim packages on the different distributions (vim-common on fedora, vim-runtime on ubuntu, etc.) when they corresponding package maintainer decides to update vim.


So to summarize: 

John, let me know if I can include the GPLv2+ license information into your files. It would then be nice if you either could write a small README file or a small text snipplet that can be include into the files regarding the usage of them.

I will then "upload" the files into the repository and let them "install" in the different release package formats under /usr/share/doc/task-<version>/examples/vim - it is then up to the user to install them into their corresponding ~/.vim

**Comment by taskwarrior** [comment]
_David Patrick on 2009-06-25T11:56:21Z says:_ 

 It's great so see such expertise applied to vim syntax colors, and I think it will improve taskwarrior immensely. 
I also see that down the road we will find an ongoing struggle to herd the color-cats. I would like to toss out an idea that is characteristically complicated, but I hope will result in a seamless, low maintenance, happy color result. 

This is early discussion, only 2/3 baked, so in no particular order.

task should come with 2 built-in color schemes (one for dark terminals, one for light) and these colors should be described (selected and manipulated) in .taskrc.

vim syntax files should not be bundled, but should be automatically generated (and installed) every time a user makes a change in the color profile.  For colors to be useful. they have to be consistent and easy to implement. The "task color" command could be augmented with "task color vim" to force the generation (and installation) of a fresh set of .vim files to match current color choices.

Parallel processes should take care of bash colorization (task color prompt) as well.

**Comment by taskwarrior** [comment]
_John Florian on 2009-06-25T12:48:02Z says:_ 

 Paul, thanks for the reminder about 'task 1 edit'.  That should be trivial.

Fredrico, yes please go ahead and add the GPLv2+ license info and I'll try to get a README written up tonight.  You also did an excellent job of summarizing the do's and don'ts of packaging.

David, as we discussed before I completely agree with the 2 built-in schemes.  I like your idea of the tighter integration with the .taskrc, but I'll have to experiment to see how possible that is.  I know when we launch vim we can specify an alternate vimrc file, but that will probably be only partly good because we'll also drop all the user's normal preferences.  Of course, we must be careful to honor the user's EDITOR setting as it may not be vim.  That doesn't mean we can't inspect it and do the right thing if it does happen to be vim.

**Comment by taskwarrior** [comment]
_David Patrick on 2009-06-25T13:49:39Z says:_ 

 John Florian wrote:
>I know when we launch vim we can specify an alternate vimrc file, but that will probably be only partly good because we'll also drop all the user's normal preferences.
rather than point to an alternate vimrc, I imagine generating a new task.vim file each time the task colors have changed. the command "task color vim" generates a new task.vim and plunks it in the (derived or  user specified in .taskrc) syntax folder

>  Of course, we must be careful to honor the user's EDITOR setting as it may not be vim.
  That doesn't mean we can't inspect it and do the right thing if it does happen to be vim.

Are you suggesting that not everyone uses vim ?? oh we'll have to change THAT ;-)

**Comment by taskwarrior** [comment]
_John Florian on 2009-06-30T20:23:39Z says:_ 

 David Patrick wrote:
>rather than point to an alternate vimrc, I imagine generating a new task.vim file each time the >task colors have changed. the command "task color vim" generates a new task.vim and plunks it in >the (derived or user specified in .taskrc) syntax folder

That's an interesting approach, but I think it's a bit naughty for one package to go diddling around in another package's config data like that.

Anyway, before we get too carried away, I need to bring up another point.  There isn't any nice way right now for vim to recognize the contents 'task 1 edit' as something from task.  Sure, I could look at the comments (which could easily change), but I think we should introduce a banner comment, something along the lines of a shebang line that indicates if this is a task edit.  It would be good to do this with {pending,completed}.data also.  Maybe it could indicate the file format version too.  Then vim can inspect the first few lines looking for this telltale banner.  I did the data files based on file name, but I suppose a user could easily have such file names unrelated to task and our colors might be rather annoying in that case.

**Comment by taskwarrior** [comment]
_David Patrick on 2009-06-30T20:56:28Z says:_ 

 John Florian wrote:
> That's an interesting approach, but I think it's a bit naughty for one package to go diddling around in another package's config data like that.

Well if .vimrc has to be modified, that would be naughty, but if it's just a matter of having entries in two of vims folders. then I don't feel to cheeky about it. 
We could keep the actual syntax file in ~/.task/syntax/task.vim, and have the command "task color vim" insert (or update) the symlink.
> 
> Anyway, before we get too carried away, I need to bring up another point.  There isn't any nice way right now for vim to recognize the contents 'task 1 edit' as something from task.  Sure, I could look at the comments (which could easily change), but I think we should introduce a banner comment, something along the lines of a shebang line that indicates if this is a task edit.  It would be good to do this with {pending,completed}.data also.  Maybe it could indicate the file format version too.  Then vim can inspect the first few lines looking for this telltale banner.  I did the data files based on file name, but I suppose a user could easily have such file names unrelated to task and our colors might be rather annoying in that case.

That's a great idea, John, as long as Paul doesn't mind changing the file format AGAIN !

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-06-30T21:32:05Z says:_ 

 I'd be happy to modify the file for "task <id> edit" in any way.  Shebang-like identifier, file "extension", or whatever.  Name it.

But as for the pending.data, completed.data, undo.data and unmentioned upcoming additional files, I'd rather not include a header.  I also don't want to change the file format, unless it's for a very good reason.  Got one?

**Comment by taskwarrior** [comment]
_John Florian on 2009-07-01T18:46:08Z says:_ 

 Paul Beckingham wrote:
> I'd be happy to modify the file for "task <id> edit" in any way.  Shebang-like identifier, file >"extension", or whatever.  Name it.
> 
> But as for the pending.data, completed.data, undo.data and unmentioned upcoming additional files, >I'd rather not include a header.  I also don't want to change the file format, unless it's for a >very good reason.  Got one?

If we can do a standard file name extension, such as .task, I think that would be the easiest.  I just now reviewed -- sorry, should have done this originally -- /usr/share/vim/vim72/filetype.vim and it definitely looks like extensions are the preferred way.  I've helped support the Progress RDBMS language in the past, which used extensions of .w, .p and .i (if any at all) and these were too common and always required peeking inside the file to guestimate ... a PITA best avoided.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-07-01T19:09:54Z says:_ 

 John, I will change "task edit" to use the .task file extension in it's temporary file name.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2009-07-01T19:34:23Z says:_ 

 Done.

**Comment by taskwarrior** [comment]
_John Florian on 2009-07-02T17:45:45Z says:_ 

 Here's an updated tarball that now provides color highlighting for 'task 1 edit', cleaned up support files for the already supported {pending,complete}.data files, an installation README and GPL licensing for all.

I'm thinking this is good shape now for sending to Bram so that it can go into VIM proper.  So, I'll begin pursuing that job.

**Comment by taskwarrior** [comment]
_Federico Hernandez on 2009-07-02T18:36:26Z says:_ 

 > Here's an updated tarball that now provides color highlighting for 'task 1 edit', cleaned up support files for the already supported {pending,complete}.data files, an installation README and GPL licensing for all.

Downloaded and will be pushed later on.


> I'm thinking this is good shape now for sending to Bram so that it can go into VIM proper.  So, I'll begin pursuing that job.

Good luck. If it helps, you know that task is part of Fedora and on its way at some point into Ubuntu/Debian.

**Comment by pbeckingham** [comment]
[TW-1191_task-vim-color.tar.gz](https://github.com/GothenburgBitFactory/taskwarrior/files/1756302/TW-1191_task-vim-color.tar.gz)
